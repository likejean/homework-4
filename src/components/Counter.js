import React, {useState} from 'react';
import { Button } from 'react-materialize';


export default ({ buttonMap }) => {
    const [counter, setCounter] = useState(0);

    const handleButtonClick = e => {
        let step = parseInt(e.target.getAttribute('step'));
        setCounter(counter + step);
    }
    const handleResetClick = e => setCounter(0);

    return (
        <div className="container">
            <div className="row counter">
                {buttonMap.lower.map(el => <Button className="col-5 z-depth-3" onClick={handleButtonClick} step={el}
                                          name="subtract" key={el}>{el}</Button>)}
                <div className="counter-container row">
                    <h2 className="col-2 counter-label">Counter: {counter}</h2>
                    <Button
                        id="reset-button"
                        large
                        node="a"
                        style={{
                            margin: 5
                        }}
                        waves="light"
                        type="submit"
                        onClick={handleResetClick}
                    >
                        RESET COUNTER
                    </Button>
                </div>
                {buttonMap.upper.map(el => <Button className="col-5 z-depth-3" onClick={handleButtonClick} step={el}
                                          name="add" key={el}>{el}</Button>)}
            </div>
        </div>
    )
}