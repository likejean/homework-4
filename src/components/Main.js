import React, { useState } from 'react';
import Range from './Range';
import Counter from './Counter';

export default () => {
    const [ rangeLimits, setRangeLimits ] = useState({
            lower : 1,
            upper : 1,
            errors: {
                lower: "",
                upper: ""
            },
            fieldStatus: {
                upper: false,
                lower: false
            }
        });
    const [ buttonMap, setButtonMap ] = useState([1]);
    const [ error, setError ] = useState('');
    const [ resetClick, setResetClick ] = useState(false);

    const handleRangeChange = e => {
        const { name, value } = e.target;
        let errors = rangeLimits.errors;
        let fieldStatus = rangeLimits.fieldStatus;
        setRangeLimits({ ...rangeLimits, [ name ]: value });
        let tempUpper = rangeLimits[ "upper" ];
        let tempLower = rangeLimits[ "lower" ];

        //User Error Validation

        switch (name) {
            case 'upper':
                errors.upper =
                    isNaN(value) || Number.isInteger(value)
                    ? 'ERROR! Input must be an INTEGER'
                    : parseInt(value) >= 10
                        ? 'ERROR! Must be less than 10'
                        : parseInt(value) < tempLower
                            ? 'ERROR! Upper Limit Must be GREATER than Lower Limit'
                            : ''
                ;
                fieldStatus.lower =
                        parseInt(value) < tempLower
                        ? true
                        : false
                ;

                break;
            case 'lower':
                errors.lower =
                    isNaN(value) || Number.isInteger(value)
                    ? 'ERROR! Input must be an INTEGER'
                    : parseInt(value) < 1
                        ? 'ERROR! Must be greater than zero'
                        : parseInt(value) > tempUpper
                            ? 'ERROR! Lower Limit Must be Less than Upper Limit'
                            : ''
                ;
                fieldStatus.upper =
                    parseInt(value) > tempUpper
                        ? true
                        : false
                ;

                break;
            default:
                break;
        }
    }

    const handleCounterReset = () => setResetClick(true);

    const handleSubmit = () => {
        let range = parseInt(rangeLimits.upper) - parseInt(rangeLimits.lower);
        range >= 0 ?
            setButtonMap(Array(range + 1).fill().map((__, idx) => parseInt(rangeLimits.lower) + idx))
            :
            setError('Invalid Inputs! Check if Upper Limit is Greater than Lower Limit');
    }
    const props = {
        rangeLimits: rangeLimits,
        buttonMap: buttonMap,
        error: error,
        resetClick: resetClick,
        setResetClick: setResetClick,
        handleRangeChange: handleRangeChange,
        handleSubmit: handleSubmit,
        handleCounterReset: handleCounterReset
    }
    return (
        <React.Fragment>
            <Range {...props}/>
            <Counter {...props}/>
        </React.Fragment>
    )
}