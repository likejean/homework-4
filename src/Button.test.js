import React from 'react';
import Enzyme, { shallow, render, mount } from "enzyme";
import { createSerializer } from 'enzyme-to-json'
import Main from './components/Main';
import Counter from './components/Counter';
import sinon from 'sinon';
import Adapter from "enzyme-adapter-react-16";

const props = {
    spy: sinon.spy(),
    buttonMap: {lower: [-1], upper: [1]}
};

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));


describe("Counter Component", () => {


    it('renders hr element', () => {
        const wrapper = shallow(<Main />);
        expect(wrapper.find('hr').length).toEqual(1);
        expect(wrapper.find('h1').length).toEqual(1);
    });

    it('renders the html of the title', () => {
        const wrapper = shallow(<Main />);
        const expectedOutput =
            '<h1 style="text-align:center;color:#26a69a"></h1>';
        const realOutput = wrapper.find('h1').html();
        expect(realOutput.indexOf(expectedOutput) > -1).toEqual(true);
    });

    it('renders the html of the title', () => {
        const wrapper = shallow(<Main />);
        const expectedOutput =
            '<h1 style="text-align:center;color:#26a69a"></h1>';
        const realOutput = wrapper.find('h1').html();
        expect(realOutput.indexOf(expectedOutput) > -1).toEqual(true);
    });

    it('renders Counter Component correctly', () => {
        const wrapper = shallow(
            <Counter { ...props } />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('outputs html markup result correctly', () => {
        const wrapper = render(
            <Counter { ...props } />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('outputs html markup result correctly', () => {
        const wrapper = render(
            <Counter { ...props } />
        );
        const text = wrapper.find('h2.counter-label').text();
        expect(text).toBe('Counter: 0');
    });

    it('calls "handleCounterReset" on button click', () => {
        const wrapper = shallow(
            <Counter { ...props } />
        );
        expect(wrapper).toMatchSnapshot();

        wrapper.find('#reset-button').simulate('click');
        expect(props.spy.calledOnce).toBe(true);
    });

});


