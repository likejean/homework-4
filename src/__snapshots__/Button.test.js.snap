// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component calls "handleCounterReset" on button click 1`] = `
<div
  className="container"
>
  <div
    className="row counter"
  >
    <Button
      className="col-5 z-depth-3"
      key="-1"
      name="subtract"
      node="button"
      onClick={[Function]}
      step={-1}
    >
      -1
    </Button>
    <div
      className="counter-container row"
    >
      <h2
        className="col-2 counter-label"
      >
        Counter: 
        0
      </h2>
      <Button
        id="reset-button"
        large={true}
        node="a"
        onClick={[Function]}
        style={
          Object {
            "margin": 5,
          }
        }
        type="submit"
        waves="light"
      >
        RESET COUNTER
      </Button>
    </div>
    <Button
      className="col-5 z-depth-3"
      key="1"
      name="add"
      node="button"
      onClick={[Function]}
      step={1}
    >
      1
    </Button>
  </div>
</div>
`;

exports[`Counter Component outputs html markup result correctly 1`] = `
<div
  class="container"
>
  <div
    class="row counter"
  >
    <button
      class="btn col-5 z-depth-3"
      name="subtract"
      step="-1"
    >
      -1
    </button>
    <div
      class="counter-container row"
    >
      <h2
        class="col-2 counter-label"
      >
        Counter: 0
      </h2>
      <a
        class="waves-effect waves-light btn-large"
        id="reset-button"
        style="margin:5px"
        type="submit"
      >
        RESET COUNTER
      </a>
    </div>
    <button
      class="btn col-5 z-depth-3"
      name="add"
      step="1"
    >
      1
    </button>
  </div>
</div>
`;

exports[`Counter Component renders Counter Component correctly 1`] = `
<div
  className="container"
>
  <div
    className="row counter"
  >
    <Button
      className="col-5 z-depth-3"
      key="-1"
      name="subtract"
      node="button"
      onClick={[Function]}
      step={-1}
    >
      -1
    </Button>
    <div
      className="counter-container row"
    >
      <h2
        className="col-2 counter-label"
      >
        Counter: 
        0
      </h2>
      <Button
        id="reset-button"
        large={true}
        node="a"
        onClick={[Function]}
        style={
          Object {
            "margin": 5,
          }
        }
        type="submit"
        waves="light"
      >
        RESET COUNTER
      </Button>
    </div>
    <Button
      className="col-5 z-depth-3"
      key="1"
      name="add"
      node="button"
      onClick={[Function]}
      step={1}
    >
      1
    </Button>
  </div>
</div>
`;
